spec/integration/client_options_spec.rb

```diff
@@ -0,0 +1,441 @@
+    it_behaves_like 'an auth mechanism with ssl'
+    it_behaves_like 'an auth mechanism that doesn\'t support auth_mech_properties'
+    it_behaves_like 'an auth mechanism that doesn\'t support invalid auth sources'
+
+    context 'with URI options' do
+      let(:credentials) { "#{user}@" }
+      let(:options) { '?authMechanism=MONGODB-X509' }
+
+      it 'sets default auth source' do
+        expect(client.options[:auth_source]).to eq('$external')
+      end
+
+      context 'when username is not provided' do
+        let(:credentials) { '' }
+
+        it 'recognizes the mechanism with no username' do
+          expect(client.options[:user]).to be_nil
+        end
+      end
+
+      context 'when a password is provided' do
+        let(:credentials) { "#{user}:password@" }
+
+        it 'throws an error on client creation' do
+          expect {
+            client
+          }.to raise_error(Mongo::Auth::InvalidConfiguration, /password is not supported/)
+        end
+      end
+    end
+
+    context 'with client options' do
+      let(:client_opts) { { auth_mech: :mongodb_x509, user: user } }
+
+      # TODO: get this test passing
+      # it 'sets default auth source' do
+      #   expect(client.options[:auth_source]).to eq('$external')
+      # end
+
+      context 'when username is not provided' do
+        let(:client_opts) { { auth_mech: :mongodb_x509} }
+
+        it 'recognizes the mechanism with no username' do
+          expect(client.options[:user]).to be_nil
+        end
+      end
+
+      context 'when a password is provided' do
+        let(:client_opts) { { auth_mech: :mongodb_x509, user: user, password: 'password' } }
+
+        it 'throws an error on client creation' do
+          expect {
+            client
+          }.to raise_error(Mongo::Auth::InvalidConfiguration, /password is not supported/)
+        end
+      end
+    end
+  end
+
+  context 'with no auth mechanism provided' do
+    context 'with URI options' do
+      context 'with no credentials' do
+        it 'creates a client with epty credentials' do
+          expect(client.options[:user]).to be_nil
+          expect(client.options[:password]).to be_nil
+        end
+      end
+
+      context 'with empty credentials' do
+        let(:credentials) { '@' }
+
+        it 'does not allow a client to be created with no username' do
+          expect {
+            client
+          }.to raise_error(Mongo::Auth::InvalidConfiguration, /empty username is not supported/)
+        end
+      end
+    end
+
+    context 'with client options' do
+      context 'with no credentials' do
+        it 'creates a client with empty credentials' do
+          expect(client.options[:user]).to be_nil
+          expect(client.options[:password]).to be_nil
+        end
+      end
+
+      context 'with empty credentials' do
+        let(:client_opts) { { user: '', password: '' } }
+
+        it 'does not allow a client to be created with no username' do
```

Comment:
Either "with empty username" or just say "does not allow a client to be created". An empty string is different from no input.